torch/_inductor/codecache.py

```diff
@@ -1822,16 +1832,101 @@ def load(cls, source_code: str) -> CDLL:
 
         return cls.cache[key]
 
 
+class CppPythonBindingsCodeCache(CppCodeCache):
+    cache: Dict[str, Union[CDLL, ModuleType]] = {}
+    clear = staticmethod(cache.clear)
+    cpp_compile_command_flags = {
+        "include_pytorch": True,
+        "shared": True,
+    }
+    suffix_template = textwrap.dedent(
+        """
+        // Python bindings to call kernel():
+        #define PY_SSIZE_T_CLEAN
+        #include <Python.h>
+
+        // This is defined in guards.cpp so we don't need to import PyTorch headers that are slooow
+        extern "C" void* _torchinductor_pyobject_tensor_data_ptr(PyObject* obj);
+
+        template <typename T> static inline T parse_arg(PyObject* args, size_t n) {
+            static_assert(std::is_pointer<T>::value, "arg type must be pointer or long");
+            return static_cast<T>(_torchinductor_pyobject_tensor_data_ptr(PyTuple_GET_ITEM(args, n)));
+        }
+        template <> inline long parse_arg<long>(PyObject* args, size_t n) {
+            auto result = PyLong_AsSsize_t(PyTuple_GET_ITEM(args, n));
+            if(result == -1 && PyErr_Occurred())
+                [[unlikely]] throw std::runtime_error("expected int arg");
+            return result;
+        }
+
+        static PyObject* kernel_py(PyObject* self, PyObject* args) {
+            try {
+                if(!PyTuple_CheckExact(args))
+                    [[unlikely]] throw std::runtime_error("tuple args required");
+                if(PyTuple_GET_SIZE(args) != %s)
+                    [[unlikely]] throw std::runtime_error("requires %s args");
+                kernel(%s);
+                Py_RETURN_NONE;
+            } catch(std::exception const& e) {
+                PyErr_SetString(PyExc_RuntimeError, e.what());
+                return nullptr;
+            }
+        }
+
+        static PyMethodDef py_methods[] = {
+            {"kernel", kernel_py, METH_VARARGS, ""},
+            {NULL, NULL, 0, NULL}};
+
+        static struct PyModuleDef py_module =
+            {PyModuleDef_HEAD_INIT, "kernel", NULL, -1, py_methods};
+
+        PyMODINIT_FUNC PyInit_kernel(void) {
+            return PyModule_Create(&py_module);
+        }
+        """
+    )
+
+    @classmethod
+    def _load_library_inner(cls, path: str, key: str) -> ModuleType:
+        return importlib.machinery.ExtensionFileLoader(
+            f"{key}.kernel", path
+        ).load_module()  # type: ignore[call-arg]
+
+    @classmethod
+    def load_pybinding(cls, argtypes: List[str], source_code: str) -> Any:
+        """
+        Wrap a C++ function in fast Python bindings.
+
+        Args:
+            argtypes: The types of args to kernel(), e.g. ["float*", "long"]
+            source_code: C++ source code containing a kernel() function
+
+        Returns:
+            A python version of kernel()
+        """
+        parseargs = ", ".join(
+            f"parse_arg<{argtype.replace('const ', '')}>(args, {n})"
```

Comment:
Why dropping `const` here?

Reply:
It would require defining two templates `parse_arg<long>` and `parse_arg<const long>`.